### PR TITLE
Limit the rate at which logger errors are logged into daemon logs

### PR DESCRIPTION
**Pull request by angelcar**
Logging to daemon logs every time there's an error with a log driver can be
problematic since daemon logs can grow rapidly, potentially exhausting disk
space.

Instead, it's preferable to limit the rate at which log driver errors are allowed
to be written. By default, this limit is 333 entries per second max.

Signed-off-by: Angel Velazquez <angelcar@amazon.com>

**- What I did**
Created a package-level `LogDriverError` function that limits the rate at which errors produced by log drivers can be written to daemon logs. 

**- How I did it**

Used `golang.org/x/time/rate` Limiter.

**- How to verify it**
* Verified the solution actually works by building my Docker changes in an EC2 box, and manually checked that the logging rate was limited as expected.

**- Description for the changelog**

Logging errors produced by log drivers are now limited to 333 per second max. This changes the previous behavior where each container log error was also logged in the daemon logs.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by IRCody** [comment]
Some questions I had while reading this:

* Is this problem unique to cloudwatchlogs?

It seems like something could be done about this inside ring.go if it's not specific to cloudwatchlogs. If it is specific it might be interesting to explore why. Are we doing something improper in cloudwatchlogs that makes this uniquely necessary? 

* Is making the error that's going per-log message that is discarded debug instead of error reasonable?
*  Were there any other alternative methods considered?

File: daemon/logger/awslogs/cloudwatchlogs.go [hunk 1]
```diff
@@ -417,7 +423,8 @@ func (l *logStream) Log(msg *logger.Message) error {
 		case l.messages <- msg:
 			return nil
 		default:
-			return errors.New("awslogs buffer is full")
+			atomic.AddInt32(&l.discardedMessages, 1)
```
Comment (FFMMM) [inline daemon/logger/awslogs/cloudwatchlogs.go:426]: would it be worth adding an explanation as to why logs are being discarded ? (ie `buffer full` or some other issue)

File: daemon/logger/awslogs/cloudwatchlogs.go [hunk 2]
```diff
@@ -513,6 +520,39 @@ var newTicker = func(freq time.Duration) *time.Ticker {
...
+
+// reportDiscardedMessages periodically checks discardedMessages counter, logging a warning if it's greater than zero.
+// The warning can be useful for the user to realize that container logs are being discarded. Container logs will be
+// discarded when this driver is configured in non-blocking mode and logs arrive faster than they can be processed
+// (i.e. sent to CloudWatch).
+func (l *logStream) reportDiscardedMessages() {
+	ticker := newTicker(time.Minute)
+	defer ticker.Stop()
+	for {
+		// If logger is closed we are done
```
Comment (FFMMM) [inline daemon/logger/awslogs/cloudwatchlogs.go:541]: So the logger could be closed but we could've missed a warning message to log that we did discard some logs, right? Do we care about that in this case?

Do we want to have a unit test in that scenario?

**Comment by FFMMM** [comment]
Not a moby reviewer or maintainer but added some small comments on the changes.

Overall this LGTM!

In addition to @IRCody's comments above I'd also ask if this change is "backwards incompatible" somehow. IOW, is anyone expecting an error on the other side?

**Comment by angelcar** [comment]
> Is this problem unique to cloudwatchlogs?

I think it's unique to `cloudwatchlogs` since this driver is blocking/non-blocking aware, and uses [non-blocking channel semantics](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/awslogs/cloudwatchlogs.go#L415-L421), thus discarding the message and returning an error when the channel is full. None of the other drivers seem to discard messages in this way and only return errors when there's a problem writing to the destination (e.g. error writing to disk).  

> It seems like something could be done about this inside ring.go if it's not specific to cloudwatchlogs. If it is specific it might be interesting to explore why. Are we doing something improper in cloudwatchlogs that makes this uniquely necessary?

There are two main issues with the way `cloudwatchlogs` operates. One is that it pretty much renders the ring buffer useless because `cloudwatchlogs.Log` never blocks, so the messages are immediately transferred from the ring buffer to the buffered channel. This means the channel is always the one that discards the logs, rather than the ring buffer. So, one could say that the ring buffer expects the driver's `Log` method to block.

The second problem (addressed in this PR) is that when `cloudwatchlogs` discards a message, it returns an error, which is then [logged](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/ring.go#L131) by the ring buffer. This is problematic when many logs are discarded (e.g. when CloudWatch is slow, or when logging rate is just to high), as explained in the summary of this PR.

>Is making the error that's going per-log message that is discarded debug instead of error reasonable?

To be able to do that, the ring logger would need to be able to differentiate between errors. For example, cloudwatchlogs returns `"awslogs buffer is full"` [error](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/awslogs/cloudwatchlogs.go#L420). So wither ring buffer would need to be aware of specific errors returned by different drivers, which doesn't feel right, or there would need to be a `MssageDiscarded` error type which would then become some sort of API error; this is also weird since only cloudwatchlogs discards messages, as mentioned before.

Ultimately, seeing `"awslogs buffer is full"` logged thousands of times doesn't seem to add much value but it does create risk of disk exhaustion, so I don't think it makes much sense to log every time a message is discarded. 
 
> Were there any other alternative methods considered?

Yes, I considered to eliminate the non-blocking channel writes from `cloudwatchlogs.Log` and just let it block, similar to what the [splunk](https://github.com/moby/moby/blob/68bec0fcf7a5eeb59c027287d06598098edc9f2c/daemon/logger/splunk/splunk.go#L405) driver does, but at this point I think that would be a backwards incompatible change that can break someone,  for example, the `awslogs-max-buffered-events` log option was added a few years ago in order to increase the size of the buffered channel, which suggest people are aware of the asynchronous nature of `cloudwatchlogs.Log`.

Another option was to do the exact same thing I am doing in this PR, but inside ring logger, but like I said before, ring buffer would need to be able to differentiate when an error is really a message discarded, which seems incorrect since only cloudwatch logs discards messages.

File: daemon/logger/awslogs/cloudwatchlogs.go (hunk 1 above)
Comment (angelcar) [reply daemon/logger/awslogs/cloudwatchlogs.go:426]: Sure, I will mention that the reason is because the buffer is full in the part that logs the warning

File: daemon/logger/awslogs/cloudwatchlogs.go (hunk 2 above)
Comment (angelcar) [reply daemon/logger/awslogs/cloudwatchlogs.go:541]: Good call. I think I can just check again if there are discarded messages before returning and we should be covered since no new messages are accepted after the driver is closed.

**Comment by angelcar** [comment]
> Not a moby reviewer or maintainer but added some small comments on the changes.
> 
> Overall this LGTM!
> 
> In addition to @IRCody's comments above I'd also ask if this change is "backwards incompatible" somehow. IOW, is anyone expecting an error on the other side?

In this case the other side would be the ring logger, which only [logs the error ](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/ring.go#L131). So this should be backwards compatible. Also, as mentioned in response @IRCody, I contend that discarding a message is even an error, since it is the expected behavior.

File: daemon/logger/awslogs/cloudwatchlogs.go (hunk 1 above)
Comment (angelcar) [reply daemon/logger/awslogs/cloudwatchlogs.go:424]: done

**Comment by thaJeztah** [comment]
> Logging every time a message is discarded can be problematic since Docker logs can grow rapidly, potentially exhausting disk space.

I haven't looked in depth into the changes/issue but are these the "container" logs, or _daemon_ logs that grow? Daemon logs are handled by the system's logs (usually systemd/journald) and would already be limited there (both in size, and (IIRC) duplicate messages are grouped/discarded).

If it's the _container_ logs, then the `local` logging driver would already have log-rotation in place. While I can see this could cause a larger amount of logs being produced, I feel that's more an issue "in general", and a bit unsure if these particular errors should be special-cased, or if the local logging driver ("ring buffer") should be improved to detect repeated messages, and discard / group them.

**Comment by angelcar** [comment]
> > Logging every time a message is discarded can be problematic since Docker logs can grow rapidly, potentially exhausting disk space.
> 
> I haven't looked in depth into the changes/issue but are these the "container" logs, or _daemon_ logs that grow? Daemon logs are handled by the system's logs (usually systemd/journald) and would already be limited there (both in size, and (IIRC) duplicate messages are grouped/discarded).
> 

Hi, thanks for looking into this. The logs that grow are the daemon logs.
Unfortunately, not everyone uses systemd/journald to manage the daemon, and different people might have different log configurations. (e.g. Windows, Ubuntu 14.10-)

> If it's the _container_ logs, then the `local` logging driver would already have log-rotation in place. While I can see this could cause a larger amount of logs being produced, I feel that's more an issue "in general", and a bit unsure if these particular errors should be special-cased, or if the local logging driver ("ring buffer") should be improved to detect repeated messages, and discard / group them.

The particular log statement in question is [this](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/ring.go#L131), in the ring.go file. This statement logs the error that the driver's `Log` method returns. For actual errors writing to the destination, or writing after the logger is closed, this makes sense, but discarded messages due to buffer being full are not errors since that's the expected behavior, IMO. In fact, the ring logger itself doesn't return an error when a message is discarded when the buffer is full, as seen [here](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/ring.go#L177), this is in contrast with what cloudwatch does, as seen [here](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/awslogs/cloudwatchlogs.go#L420)

Thoughts?

**Comment by cpuguy83** [comment]
It seems like it would be better to put something like this in the log copier where we actually do the logging.

**Comment by angelcar** [comment]
> It seems like it would be better to put something like this in the log copier where we actually do the logging.

Thanks for looking into this. So, the log statement motivating this PR is not in copier, but in ring.go, as described below:
* When awslogs is in `non-blocking` mode and the internal channel gets full, `awslogs.Log()` discards the message and returns an error, as seen [here](https://github.com/moby/moby/blob/master/daemon/logger/awslogs/cloudwatchlogs.go#L420)
* The error is then logged by `ring.run()` [here](https://github.com/moby/moby/blob/master/daemon/logger/ring.go#L131)
* Note that copier is not logging the error in this case because `ring.Log()` (which is what copier invokes in `non-blocking` mode) is actually not the one discarding the message.

* For log drivers other than awslogs, the ring buffer is the one that discards the message, but`ring.Log()` doesn't return error when messages are discarded, as seen [here](https://github.com/moby/moby/blob/master/daemon/logger/ring.go#L177), so there's nothing for copier to log.

Thoughts?

**Comment by cpuguy83** [comment]
@angelcar Right ok, so this is 2 places where it can happen. We need to handle the log output it both places.

**Comment by angelcar** [comment]
> @angelcar Right ok, so this is 2 places where it can happen. We need to handle the log output it both places.

Happy to do the change, but as of now this situation can't happen in copier since `ring.Log()` doesn't return "message discarded" errors. Are you suggesting that `ring.Log()` should now to return errors so that copier can handle them like I do in this PR?

**Comment by cpuguy83** [comment]
@angelcar No. I'm saying that we need to handle this in the ring logger AND we need to handle it in the log copier for the case of blocking mode.

**Comment by angelcar** [comment]
> @angelcar No. I'm saying that we need to handle this in the ring logger AND we need to handle it in the log copier for the case of blocking mode.

Could you share the rationale behind implementing this for blocking AND non-blocking mode?

I think this should probably only be done specifically for "message dropped" errors, which, as of now, are only produced by `awslogs.Log()` in non-blocking mode. 

The rationale behind only doing this for "message dropped" is:
* "message dropped" errors have the potential to flood daemon logs
* Logging "message dropped" errors individually doesn't really add much value

If we can say the same about any other error returned by any driver's `Log()` (risk of flooding, no value in logging individual errors), then we I could apply the logic in this PR to log copier, as you suggest.

**Comment by cpuguy83** [comment]
In all error cases (as I recall) the message is dropped and dumped to the daemon's log.
The cause of the error doesn't really change anything.
If an endpoint is down, or a disk is full, whatever, it doesn't do much good to spam the logs for each entry.

**Comment by angelcar** [comment]
> In all error cases (as I recall) the message is dropped and dumped to the daemon's log.
> The cause of the error doesn't really change anything.
> If an endpoint is down, or a disk is full, whatever, it doesn't do much good to spam the logs for each entry.

Alright, thanks for clarifying. Will submit a new rev accordingly.

**Comment by angelcar** [comment]
> @angelcar No. I'm saying that we need to handle this in the ring logger AND we need to handle it in the log copier for the case of blocking mode.

@cpuguy83 done!

**Comment by cpuguy83** [comment]
@angelcar Thanks for looking at this so quickly.
I was thinking we can handle this in a simpler way w/o having to write another log driver in between all everything.

Instead where we check the error in the calls to log we can pass it off like `logger.LogError(err)` which would make the decision if it should log it or not. This may be tricky depending on how state management should be worked... like should we do this globally? per container... I tend to hate global things but daemon logs are global. Would love to hear your thoughts as well as others'.

**Comment by angelcar** [comment]
> @angelcar Thanks for looking at this so quickly.
> I was thinking we can handle this in a simpler way w/o having to write another log driver in between all everything.
> 
> Instead where we check the error in the calls to log we can pass it off like `logger.LogError(err)` which would make the decision if it should log it or not. 

I am not sure if I like the approach you suggest because:
* There are many places where `logger.LogError(err)` would need to be called. 
  * Ring logger [L107](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/ring.go#L107)
  * Ring logger [L131](https://github.com/moby/moby/blob/c0c3e58bb2b3714683658d331a138b2d80a6e974/daemon/logger/ring.go#L131)
  * Copier [L130](https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/logger/copier.go#L130)
  * Copier [L163](https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/logger/copier.go#L163)
  * Potentially in other asynchronous loggers/wrappers in the future (similar to ring), which will be easy to miss

> This may be tricky depending on how state management should be worked... like should we do this globally? per container... I tend to hate global things but daemon logs are global. Would love to hear your thoughts as well as others'.

I do see value in doing this per container. A user might have multiple containers and maybe just one of them is generating log errors, whereas the others might be totally fine; as a user I'd like to be able to identify who the problem child is.

**Comment by cpuguy83** [comment]
@angelcar I'm not too worried about having to catch this in multiple places. I don't see us adding any new async logging mechanism.
I don't want to take a case that we only handle on error and move to an abstraction that adds another layer for every call.

I also think something like https://pkg.go.dev/golang.org/x/time/rate#Limiter.Allow would work pretty well here.

**Comment by angelcar** [comment]
> I don't want to take a case that we only handle on error and move to an abstraction that adds another layer for every call.

@cpuguy83 What is the concern with the additional wrapper? Is it performance or readability/maintainability?  
If it is performance, I think the hit of an additional call in the stack is negligible since the [log message](https://github.com/moby/moby/blob/master/daemon/logger/logger.go#L76) passed is already pointer (`Log(*logger.Message)`), so I guess no additional stuff escaping to the heap. I could write a benchmark to prove this.
If the main concern maintainability/readability, that's more debatable. I agree the multiple nesting is kind of confusing. 

Anyhow, for the sake of moving forward, do you want me to go with your suggestion of using a global `logger.LogError(err)` with a limiter? I could probably do something like:

```golang
func LogError(container string, logErr error) {
  // determine if we should log error to daemon logs based on `Limiter.Allow`
  // if above is true, log "logErr" to daemon logs, add "containerid" as logrus field
}
``` 

So basically this solution just controls the rate at which log errors can be logged to daemon logs. yes?

**Comment by angelcar** [comment]
Hi @cpuguy83, I went ahead with the simpler implementation which only limits the rate at which log diver errors can be logged to daemon logs.  
Thanks,
Angel.

File: daemon/logger/logger_error.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+// Rates based on journald defaults of 10,000 messages in 30s.
+// reference: https://www.freedesktop.org/software/systemd/man/journald.conf.html#RateLimitIntervalSec=
+var logErrorLimiter = rate.NewLimiter(333, 333)
+
+// LogDriverError logs errors produced by log drivers to the daemon logs. It also increments the logWritesFailedCount
+// metric.
+// Logging to the daemon logs is limited to 333 operations per second at most. If this limit is exceeded, the
+// logWritesFailedCount is still counted, but logging to the daemon logs is omitted in order to prevent disk saturation.
+func LogDriverError(loggerName, msgLine string, logErr error) {
```
Comment (samuelkarp) [inline daemon/logger/logger_error.go:16]: Can this be made private?

```suggestion
func logDriverError(loggerName, msgLine string, logErr error) {
```

**Review by samuelkarp — COMMENTED** [review]
I like this approach significantly more.  The limit seems high to me, but 333/s is still a reduction from no limit, so I don't really have an objection to it.

File: daemon/logger/logger_error.go (hunk 3 above)
Comment (angelcar) [reply daemon/logger/logger_error.go]: oops, changed. Thanks.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by samuelkarp — APPROVED** [review]
LGTM

**Comment by angelcar** [comment]
Thanks for the approvals, do we need anything else to merge?

**Comment by samuelkarp** [comment]
```
[2021-05-27T21:38:06.539Z]     --- FAIL: TestDockerSuite/TestPullWindowsImageFailsOnLinux (0.10s)
[2021-05-27T21:38:06.539Z]         docker_cli_pull_test.go:276: assertion failed: expected error to contain "no matching manifest for linux", got "\nCommand:  /usr/local/cli/docker pull mcr.microsoft.com/windows/servercore:ltsc2019\nExitCode: 1\nError:    exit status 1\nStdout:   \nStderr:   Error response from daemon: Get https://mcr.microsoft.com/v2/: dial tcp 255.255.255.255:443: connect: network is unreachable\n\n\nFailures:\nExitCode was 1 expected 0\nExpected no error"
```

Appears to be unrelated (looks like https://twitter.com/tianon/status/1398018248759648256)

**Comment by samuelkarp** [comment]
Looks like `TestDaemonHostGatewayIP` failed, which is known flaky: https://github.com/moby/moby/issues/42357

**Comment by samuelkarp** [comment]
`TestCreateServiceConfigFileMode` also failed https://github.com/moby/moby/issues/36907
